### PR TITLE
add a new interface to query format feature info

### PR DESCRIPTION
**Pull request by yiwenxue**
Re: cocos-creator/3d-tasks#8203 

TS engine: https://github.com/cocos-creator/engine/pull/9949#issue-1096192951

Move the query of format feature information to the new interface. In the old method, one can only know whether the device generally supports a feature, but the fact is that some formats support this feature, and others don't. With the new interface, one can query whether a feature is supported by a specific format or not.

Vulkan and GLES are tested with test-cases-3d. This modification introduced no extra errors.

### CODE REVIEW

File: cocos/renderer/gfx-gles2/GLES2Commands.cpp [hunk 1]
```diff
@@ -1031,7 +1020,7 @@ void cmdFuncGLES2CreateShader(GLES2Device *device, GLES2GPUShader *gpuShader) {
         gpuShader->glSamplerTextures.resize(gpuShader->samplerTextures.size());
 
         for (size_t i = 0; i < gpuShader->glSamplerTextures.size(); ++i) {
-            UniformSamplerTexture &        samplerTexture    = gpuShader->samplerTextures[i];
+            UniformSamplerTexture         &samplerTexture    = gpuShader->samplerTextures[i];
```
Comment (YunHsiao) [inline cocos/renderer/gfx-gles2/GLES2Commands.cpp:1023]: These changes are probably come from the wrong version of LLVM. Try installing 12.0 for now.

File: cocos/renderer/gfx-gles2/GLES2Device.cpp [hunk 2]
```diff
@@ -284,6 +246,174 @@ void GLES2Device::bindContext(bool bound) {
...
+        _formatFeatures[toNumber(Format::ASTC_RGBA_8X5)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_8X6)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_8X8)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_10X5)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_10X6)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_10X8)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_10X10)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_12X10)] |= compressedFeature;
+        _formatFeatures[toNumber(Format::ASTC_RGBA_12X12)] |= compressedFeature;
+        if (checkExtension("EXT_sRGB")) {
```
Comment (YunHsiao) [inline cocos/renderer/gfx-gles2/GLES2Device.cpp:398]: I think this check is redundant. The ASTC extension includes both RGBA & SRGBA variants.

File: cocos/renderer/gfx-gles2/GLES2Device.h [hunk 3]
```diff
@@ -80,6 +80,8 @@ class CC_GLES2_API GLES2Device final : public Device {
         });
     }
 
+    inline bool checkTextureExclusive(const Format &format) { return _textureExclusive[static_cast<size_t>(format)]; };
```
Comment (YunHsiao) [inline cocos/renderer/gfx-gles2/GLES2Device.h:83]:
```suggestion
    // check the specified format is texture-exclusive (no renderbuffers allowed)
    inline bool isTextureExclusive(const Format &format) { return _textureExclusive[static_cast<size_t>(format)]; };
```

File: cocos/renderer/gfx-gles3/GLES3Commands.cpp [hunk 4]
```diff
@@ -831,7 +831,7 @@ void cmdFuncGLES3CreateTexture(GLES3Device *device, GLES3GPUTexture *gpuTexture)
         return;
     }
 
-    if (gpuTexture->glSamples > 1 || hasAllFlags(TextureUsage::COLOR_ATTACHMENT | TextureUsage::DEPTH_STENCIL_ATTACHMENT, gpuTexture->usage)) {
+    if (!device->checkTextureExclusive(gpuTexture->format) && gpuTexture->glSamples > 1 || hasAllFlags(TextureUsage::COLOR_ATTACHMENT | TextureUsage::DEPTH_STENCIL_ATTACHMENT, gpuTexture->usage)) {
```
Comment (YunHsiao) [inline cocos/renderer/gfx-gles3/GLES3Commands.cpp:834]:
```suggestion
    if (!device->checkTextureExclusive(gpuTexture->format) && (gpuTexture->glSamples > 1 || hasAllFlags(TextureUsage::COLOR_ATTACHMENT | TextureUsage::DEPTH_STENCIL_ATTACHMENT, gpuTexture->usage))) {
```

File: cocos/renderer/gfx-gles3/GLES3Device.cpp [hunk 5]
```diff
@@ -282,6 +257,221 @@ void GLES3Device::bindContext(bool bound) {
...
+    _formatFeatures[toNumber(Format::SRGB8)]    = tempFeature;
+    _formatFeatures[toNumber(Format::SRGB8_A8)] = tempFeature;
+
+    _formatFeatures[toNumber(Format::R11G11B10F)] = tempFeature;
+    _formatFeatures[toNumber(Format::RGB9E5)]     = tempFeature;
+
+    _formatFeatures[toNumber(Format::DEPTH)]         = tempFeature;
+    _formatFeatures[toNumber(Format::DEPTH_STENCIL)] = tempFeature;
+
+    removeFlags(tempFeature, FormatFeature::SAMPLED_TEXTURE);
```
Comment (YunHsiao) [inline cocos/renderer/gfx-gles3/GLES3Device.cpp:294]: Don't do this here, the flags should be declared as straightforward as possible in advance, with each variable names reflecting the accurate semantics:
```cpp
    constexpr FormatFeature textureFeaturesNoFiltering = FormatFeature::RENDER_TARGET | FormatFeature::SAMPLED_TEXTURE | FormatFeature::STORAGE_TEXTURE;
    constexpr FormatFeature allTextureFeatures = textureFeaturesNoFiltering | FormatFeature::LINEAR_FILTER;
    constexpr FormatFeature allFeatures = allTextureFeatures  | FormatFeature::VERTEX_ATTRIBUTE;
```
All the procedural bitwise manipulations in the middle of format declarations will only make the code hard to follow or modify.

File: cocos/renderer/gfx-gles3/GLES3Device.cpp [hunk 6]
```diff
@@ -282,6 +257,221 @@ void GLES3Device::bindContext(bool bound) {
...
+    // tempFeature = completeFeature ^ FormatFeature::LINEAR_FILTER | FormatFeature::VERTEX_ATTRIBUTE;
+    tempFeature = completeFeature | FormatFeature::VERTEX_ATTRIBUTE;
+    removeFlags(tempFeature, FormatFeature::LINEAR_FILTER);
+
+    _formatFeatures[toNumber(Format::R16F)]    = tempFeature;
+    _formatFeatures[toNumber(Format::RG16F)]   = tempFeature;
+    _formatFeatures[toNumber(Format::RGB16F)]  = tempFeature;
+    _formatFeatures[toNumber(Format::RGBA16F)] = tempFeature;
+
+    // tempFeature = completeFeature ^ FormatFeature::SAMPLED_TEXTURE ^ FormatFeature::LINEAR_FILTER | FormatFeature::VERTEX_ATTRIBUTE;
```
Comment (YunHsiao) [inline cocos/renderer/gfx-gles3/GLES3Device.cpp:307]: Be sure to clear the redundant code before PR submission

File: cocos/renderer/gfx-metal/MTLDevice.h [hunk 7]
```diff
@@ -133,6 +135,8 @@ class CCMTLDevice final : public Device {
 
     std::vector<CCMTLSwapchain *> _swapchains;
 
+    std::array<bool, static_cast<size_t>(Format::COUNT)> _textureExclusive;
```
Comment (YunHsiao) [inline cocos/renderer/gfx-metal/MTLDevice.h:138]: redundant

File: cocos/renderer/gfx-validator/TextureValidator.cpp [hunk 8]
```diff
@@ -62,8 +60,12 @@ void TextureValidator::doInit(const TextureInfo &info) {
     _inited = true;
 
     CCASSERT(info.width && info.height && info.depth, "zero-sized texture?");
-    CCASSERT(!featureCheckMap.count(_info.format) || DeviceValidator::getInstance()->hasFeature(featureCheckMap[_info.format]), "unsupported format");
 
+    FormatFeature ff = FormatFeature::RENDER_TARGET;
+    if (info.usage == TextureUsage::SAMPLED) ff |= FormatFeature::SAMPLED_TEXTURE;
+    if (info.usage == TextureUsage::STORAGE) ff |= FormatFeature::STORAGE_TEXTURE;
+
+    CCASSERT(hasAllFlags(DeviceValidator::getInstance()->getFormatFeatures(info.format), ff), "Format not supported for the specified features");
```
Comment (YunHsiao) [inline cocos/renderer/gfx-validator/TextureValidator.cpp:68]:
```suggestion
    FormatFeature ff = FormatFeature::NONE;
    if (hasAnyFlags(info.usage, TextureUsage::COLOR_ATTACHMENT | TextureUsage::DEPTH_STENCIL_ATTACHMENT)) ff |= FormatFeature::RENDER_TARGET;
    if (hasFlag(info.usage, TextureUsage::SAMPLED)) ff |= FormatFeature::SAMPLED_TEXTURE;
    if (hasFlag(info.usage, TextureUsage::STORAGE)) ff |= FormatFeature::STORAGE_TEXTURE;
    
    if (ff) CCASSERT(hasAllFlags(DeviceValidator::getInstance()->getFormatFeatures(info.format), ff), "Format not supported for the specified features");
```

File: cocos/renderer/gfx-vulkan/VKDevice.cpp [hunk 9]
```diff
@@ -40,11 +40,13 @@
 #include "VKSwapchain.h"
 #include "VKTexture.h"
 #include "VKUtils.h"
+#include "gfx-base/GFXDef-common.h"
 #include "states/VKGlobalBarrier.h"
 #include "states/VKSampler.h"
 #include "states/VKTextureBarrier.h"
 
 #include "gfx-base/SPIRVUtils.h"
+#include "vulkan/vulkan_core.h"
```
Comment (YunHsiao) [inline cocos/renderer/gfx-vulkan/VKDevice.cpp:49]: redundant

File: cocos/renderer/gfx-vulkan/VKDevice.cpp [hunk 10]
```diff
@@ -40,11 +40,13 @@
 #include "VKSwapchain.h"
 #include "VKTexture.h"
 #include "VKUtils.h"
+#include "gfx-base/GFXDef-common.h"
```
Comment (YunHsiao) [inline cocos/renderer/gfx-vulkan/VKDevice.cpp:43]: redundant

File: cocos/renderer/gfx-vulkan/VKDevice.h [hunk 11]
```diff
@@ -152,6 +154,8 @@ class CC_VULKAN_API CCVKDevice final : public Device {
 
     vector<const char *> _layers;
     vector<const char *> _extensions;
+
+    std::array<bool, static_cast<size_t>(Format::COUNT)> _textureExclusive;
```
Comment (YunHsiao) [inline cocos/renderer/gfx-vulkan/VKDevice.h:158]: redundant

File: cocos/renderer/pipeline/RenderPipeline.cpp [hunk 12]
```diff
@@ -292,7 +292,7 @@ void RenderPipeline::generateConstantMacros() {
 #define CC_PLATFORM_ANDROID_AND_WEBGL 0
 #define CC_ENABLE_WEBGL_HIGHP_STRUCT_VALUES 0
         )",
-        _device->hasFeature(gfx::Feature::TEXTURE_FLOAT) ? 1 : 0,
+        (_device->getFormatFeatures(gfx::Format::RGBA32F) & gfx::FormatFeature::RENDER_TARGET) != gfx::FormatFeature::NONE ? 1 : 0,
```
Comment (YunHsiao) [inline cocos/renderer/pipeline/RenderPipeline.cpp:295]:
```suggestion
        hasAnyFlags(_device->getFormatFeatures(gfx::Format::RGBA32F), gfx::FormatFeature::RENDER_TARGET | gfx::FormatFeature::SAMPLED_TEXTURE),
```

File: cocos/renderer/pipeline/deferred/DeferredPipeline.cpp [hunk 13]
```diff
@@ -150,7 +150,7 @@ bool DeferredPipeline::activeRenderer(gfx::Swapchain *swapchain) {
 
     // update global defines when all states initialized.
     _macros.setValue("CC_USE_HDR", static_cast<bool>(sharedData->isHDR));
-    _macros.setValue("CC_SUPPORT_FLOAT_TEXTURE", _device->hasFeature(gfx::Feature::TEXTURE_FLOAT));
+    _macros.setValue("CC_SUPPORT_FLOAT_TEXTURE", (_device->getFormatFeatures(gfx::Format::RGBA32F) & (gfx::FormatFeature::RENDER_TARGET | gfx::FormatFeature::SAMPLED_TEXTURE)) != gfx::FormatFeature::NONE);
```
Comment (YunHsiao) [inline cocos/renderer/pipeline/deferred/DeferredPipeline.cpp:153]:
```suggestion
    _macros.setValue("CC_SUPPORT_FLOAT_TEXTURE", hasAnyFlags(_device->getFormatFeatures(gfx::Format::RGBA32F), gfx::FormatFeature::RENDER_TARGET | gfx::FormatFeature::SAMPLED_TEXTURE));
```

File: cocos/renderer/pipeline/forward/ForwardPipeline.cpp [hunk 14]
```diff
@@ -136,7 +136,7 @@ bool ForwardPipeline::activeRenderer(gfx::Swapchain *swapchain) {
 
     // update global defines when all states initialized.
     _macros.setValue("CC_USE_HDR", static_cast<bool>(sharedData->isHDR));
-    _macros.setValue("CC_SUPPORT_FLOAT_TEXTURE", _device->hasFeature(gfx::Feature::TEXTURE_FLOAT));
+    _macros.setValue("CC_SUPPORT_FLOAT_TEXTURE", (_device->getFormatFeatures(gfx::Format::RGBA32F) & (gfx::FormatFeature::RENDER_TARGET | gfx::FormatFeature::SAMPLED_TEXTURE)) != gfx::FormatFeature::NONE);
```
Comment (YunHsiao) [inline cocos/renderer/pipeline/forward/ForwardPipeline.cpp:139]:
```suggestion
    _macros.setValue("CC_SUPPORT_FLOAT_TEXTURE", hasAnyFlags(_device->getFormatFeatures(gfx::Format::RGBA32F), gfx::FormatFeature::RENDER_TARGET | gfx::FormatFeature::SAMPLED_TEXTURE));
```

**Review by YunHsiao — CHANGES_REQUESTED** [review]

**Comment by YunHsiao** [comment]
Also remember to change the PR title to a more accurate description of the content

File: cocos/renderer/gfx-base/GFXDef-common.h [hunk 15]
```diff
@@ -469,6 +456,17 @@ enum class TextureFlagBit : uint32_t {
 using TextureFlags = TextureFlagBit;
 CC_ENUM_BITWISE_OPERATORS(TextureFlagBit);
 
+enum class FormatFeatureBit : uint32_t {
+    NONE             = 0,
+    RENDER_TARGET    = 0x1,  // Allow usages as render pass attachments
+    SAMPLED_TEXTURE  = 0x2,  // Allow sampled reads in shaders
+    LINEAR_FILTER    = 0x4,  // Allow linear filtering when sampling in shaders or bliting
```
Comment (YunHsiao) [inline cocos/renderer/gfx-base/GFXDef-common.h:463]:
```suggestion
    LINEAR_FILTER    = 0x4,  // Allow linear filtering when sampling in shaders or blitting
```

File: cocos/renderer/gfx-validator/DescriptorSetValidator.cpp [hunk 16]
```diff
@@ -71,6 +71,29 @@ void DescriptorSetValidator::doDestroy() {
...
+        if (texture == nullptr || sampler == nullptr) continue;
+        format  = texture->getInfo().format;
+
+        if (sampler->getInfo().magFilter == Filter::LINEAR ||
+            sampler->getInfo().mipFilter == Filter::LINEAR ||
+            sampler->getInfo().minFilter == Filter::LINEAR) {
+
+            auto feature = DeviceValidator::getInstance()->getFormatFeatures(format);
+            if (!hasFlag(DeviceValidator::getInstance()->getFormatFeatures(format), FormatFeature::LINEAR_FILTER)) {
+                CC_LOG_DEBUG("[WARNING]: Format doesn't support linear filter.");
```
Comment (YunHsiao) [inline cocos/renderer/gfx-validator/DescriptorSetValidator.cpp:92]: Just use `CC_LOG_WARNING`

File: cocos/renderer/gfx-validator/InputAssemblerValidator.cpp [hunk 17]
```diff
@@ -48,6 +48,11 @@ void InputAssemblerValidator::doInit(const InputAssemblerInfo &info) {
     CCASSERT(!isInited(), "initializing twice?");
     _inited = true;
 
+    // vertex attributes validations
+    for (auto const &attribute : info.attributes) {
+        CCASSERT(hasAllFlags(DeviceValidator::getInstance()->getFormatFeatures(attribute.format), FormatFeature::VERTEX_ATTRIBUTE), "Format not supported for the specified features");
```
Comment (YunHsiao) [inline cocos/renderer/gfx-validator/InputAssemblerValidator.cpp:53]: when test against a single flag, use the `hasFlag` function instead

File: cocos/renderer/gfx-vulkan/VKDevice.cpp [hunk 18]
```diff
@@ -227,20 +228,32 @@ bool CCVKDevice::doInit(const DeviceInfo & /*info*/) {
...
+    _features[toNumber(Feature::MULTIPLE_RENDER_TARGETS)]  = true;
+    _features[toNumber(Feature::BLEND_MINMAX)]             = true;
+    _features[toNumber(Feature::COMPUTE_SHADER)]           = true;
+    _features[toNumber(Feature::INPUT_ATTACHMENT_BENEFIT)] = true;
+
+    initFormatFeature();
+
+    String compressedFmts;
+
+    if ((_formatFeatures[toNumber(Format::BC1_SRGB_ALPHA)] & FormatFeature::SAMPLED_TEXTURE) != FormatFeature::NONE) {
```
Comment (YunHsiao) [inline cocos/renderer/gfx-vulkan/VKDevice.cpp:242]: make more use of `hasFlags` instead of doing it manually

File: cocos/renderer/gfx-base/GFXDevice.h [hunk 19]
```diff
@@ -105,6 +105,7 @@ class CC_DLL Device : public Object {
     inline const String &    getRenderer() const { return _renderer; }
     inline const String &    getVendor() const { return _vendor; }
     inline bool              hasFeature(Feature feature) const { return _features[static_cast<uint32_t>(feature)]; }
+    inline FormatFeature     getFormatFeatures(Format format) { return _formatFeatures[static_cast<uint32_t>(format)]; }
```
Comment (YunHsiao) [inline cocos/renderer/gfx-base/GFXDevice.h:108]:
```suggestion
    inline bool              hasFeature(Feature feature) const { return _features[toNumber(feature)]; }
    inline FormatFeature     getFormatFeatures(Format format) { return _formatFeatures[toNumber(format)]; }
```

**Review by YunHsiao — APPROVED** [review]

File: cocos/renderer/gfx-base/GFXDef-common.h [hunk 20]
```diff
@@ -139,20 +139,6 @@ enum class SurfaceTransform : uint32_t {
...
-    TEXTURE_FLOAT_LINEAR,
-    TEXTURE_HALF_FLOAT_LINEAR,
-    FORMAT_R11G11B10F,
-    FORMAT_SRGB,
-    FORMAT_ETC1,
-    FORMAT_ETC2,
-    FORMAT_DXT,
-    FORMAT_PVRTC,
-    FORMAT_ASTC,
-    FORMAT_RGB8,
```
Comment (minggo) [inline cocos/renderer/gfx-base/GFXDef-common.h:155]: will it break compatibility?

Comment (YunHsiao) [reply cocos/renderer/gfx-base/GFXDef-common.h:155]: All references have been updated, but yes, this is a minor breaking change

**Review by minggo — APPROVED** [review]
